cmd/example-mysql/main.go

```diff
@@ -69,6 +72,35 @@ func main() {
 		}
 	})
 
+	http.HandleFunc("GET /tile/{level}/{index...}", func(w http.ResponseWriter, r *http.Request) {
+		level, index, err := parseTileLevelIndex(r.PathValue("level"), r.PathValue("index"))
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			if _, werr := w.Write([]byte(err.Error())); werr != nil {
+				klog.Errorf("/tile/{level}/{index...}: %v", werr)
+			}
+			return
+		}
+
+		tile, err := storage.ReadTile(r.Context(), level, index)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				w.WriteHeader(http.StatusNotFound)
+				return
+			}
+
+			klog.Errorf("/tile/{level}/{index...}: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Cache-Control", "public, max-age=31536000, immutable")
```

Comment:
This could be dangerous here; if someone requests a partial tile without the partial index, you will return the partial tile. When this tile grows and eventually becomes full, someone going via the cache will not be able to see it.

I think we need to look at only setting this header when the number of leaves exactly matches the tile that was requested (i.e. a partial tile was requested and exactly that many leaves were present, or a full tile was requested and we returned 256 leaves).

This should be calculable because we know that the hashes are 32 bytes long, so we can determine here how many hashes are in the returned tile. You'll have to make the path parser pull out the partial tile size though and return it (I would consider making it return the number of expected hashes, so return 256 if there is no partial specification).

Reply:
Good catch. PTAL, thanks!